src/main/java/me/zeroeightsix/kami/manager/mangers/WaypointManager.kt

```diff
@@ -1,25 +1,152 @@
+     * LinkedHashSet for all waypoints
+     * Since LinkedHashSet isn't thread safe, getting this from another thread will
+     * returns a read only copy
+     */
+    var waypoints = LinkedHashSet<Waypoint>()
+        get() {
+            return if (Thread.currentThread() != mainThread) {
+                synchronized(field) { LinkedHashSet(field) }
+            } else field
+        }
 
     /**
      * Reads waypoints from KAMIBlueWaypoints.json into the waypoints ArrayList
      */
     fun loadWaypoints(): Boolean {
-        return Waypoint.readFileToMemory()
+        /* backwards compatibility for older configs */
+        val localFile = if (legacyFormat()) oldFile else file
+        val success = try {
+            waypoints = gson.fromJson(FileReader(localFile), object : TypeToken<LinkedHashSet<Waypoint>?>() {}.type)
+            KamiMod.log.info("Waypoint loaded")
+            if (legacyFormat()) oldFile.delete()
+            true
+        } catch (e: FileNotFoundException) {
+            KamiMod.log.warn("Could not find file $configName, clearing the waypoints list")
+            waypoints.clear()
+            false
+        } catch (e: IllegalStateException) {
+            KamiMod.log.warn("$configName is empty!")
+            waypoints.clear()
+            false
+        }
+        KamiMod.EVENT_BUS.post(WaypointUpdateEvent(WaypointUpdateEvent.Type.CLEAR, null))
+        return success
     }
 
     /**
      * Saves waypoints from the waypoints ArrayList into KAMIBlueWaypoints.json
      */
     fun saveWaypoints(): Boolean {
-        return Waypoint.writeMemoryToFile()
+        return try {
+            val fileWriter = FileWriter(file, false)
+            gson.toJson(waypoints, fileWriter)
+            fileWriter.flush()
+            fileWriter.close()
+            KamiMod.log.info("Waypoint saved")
+            true
+        } catch (e: IOException) {
+            KamiMod.log.info("Failed saving waypoint")
+            e.printStackTrace()
+            false
+        }
+    }
+
+    /**
+     * file deletion does not work on OSX, issue #1044
+     * because of this, we must also check if they've used the new format
+     */
+    private fun legacyFormat(): Boolean {
+        return oldFile.exists() && !file.exists()
+    }
+
+    fun get(id: String): Waypoint? {
+        val waypoint = waypoints.firstOrNull { it.id.toString() == id }
+        KamiMod.EVENT_BUS.post(WaypointUpdateEvent(WaypointUpdateEvent.Type.GET, waypoint))
+        return waypoint
+    }
+
+    fun get(pos: BlockPos, currentDimension: Boolean = false): Waypoint? {
+        val waypoint = waypoints.firstOrNull { (if (currentDimension) it.currentPos() else it.pos) == pos }
+        KamiMod.EVENT_BUS.post(WaypointUpdateEvent(WaypointUpdateEvent.Type.GET, waypoint))
+        return waypoint
+    }
+
+    fun add(locationName: String): Waypoint {
+        val coords = Wrapper.player?.positionVector?.toBlockPos() ?: BlockPos(0, -6969, 0) // This shouldn't happen
```

Comment:
We should have better error handling then just adding a fake waypoint. Just return and KamiMod.log.warn if it won't ever happen (for better troubleshooting!!)

Reply:
Because it must return a waypoint so I made it adds a fake waypoint